pkg/cmd/server/crypto/crypto.go

```diff
@@ -142,15 +143,78 @@ var (
 )
 
 type CA struct {
+	Config *TLSCertificateConfig
+
+	SerialGenerator SerialGenerator
+}
+
+// SerialGenerator is an interface for getting a serial number for the cert.  It MUST be thread-safe.
+type SerialGenerator interface {
+	Next(template *x509.Certificate) (int64, error)
+}
+
+// SerialFileGenerator returns a unique, monotonically increasing serial number and ensures the CA on disk records that value.
+type SerialFileGenerator struct {
 	SerialFile string
-	Config     *TLSCertificateConfig
 
 	// lock guards access to the Serial field
 	lock   sync.Mutex
 	Serial int64
 }
 
+func NewSerialFileGenerator(serialFile string, createIfNeeded bool) (*SerialFileGenerator, error) {
+	// read serial file
+	var serial int64
+	serialData, err := ioutil.ReadFile(serialFile)
+	if err == nil {
+		serial, _ = strconv.ParseInt(string(serialData), 16, 64)
+	}
+	if os.IsNotExist(err) && createIfNeeded {
+		if err := ioutil.WriteFile(serialFile, []byte("0"), 0644); err != nil {
```

Comment:
> 00

Copy/paste.  I'm assuming this doesn't cause us any issues we need to deal with for legacy files?